[Weave] TypeScript coverage for media logging (#2001)
wandb/docs: weave/guides/core-types/media.mdx

description: "Log media returned in your traces, such as images and videos."
---

> W&B Weave supports logging and has dedicated displays for numerous content types such as videos, images, audio files, PDFs, CSVs and HTML. Each section of this guide provides a basic configuration example and a more advanced example of how to log the media type.

Reviewer: ```suggestion
W&B Weave supports logging and displaying numerous content types, with dedicated functions for displaying videos, images, audio clips, PDFs, CSV data, and HTML. This guide provides basic and advanced examples for logging and displaying each media type.
```
Not sure if this is what you actually mean. I'm also not sure that the manual TOC is necessary, since the right-hand TOC will have the same structure.

Author: nope, this is much better! Thank you

- **[Images](#log-images)** 
- **[Video](#log-video)** 
- **[Documents](#log-documents)**
- **[Audio](#log-audio)**
- **[HTML](#log-html)**

## Overview

<Tabs>
<Tab title="Python">

> Use type annotations like `Annotated[bytes, Content]` or `Annotated[str, Content]` as input or return types in your ops to log media to Weave. You can also annotate path arguments with `Annotated[str, Content]` to have Weave automatically open, detect and display the media within your trace.

Reviewer: ```suggestion
To log media to Weave, add type annotations like `Annotated[bytes, Content]` or `Annotated[str, Content]` as input or return types in your ops. If you annotate path arguments with `Annotated[str, Content]`, Weave automatically opens, detects, and displays the media within your trace.
```

> The examples in this guide use annotations. We recommend using annotations because they are the simplest way to start logging your media. For more advanced configurations, see the [Content API section](#using-the-contents-api).

Reviewer: This is out of order with the previous paragraph, I suspect.

</Tab>
<Tab title="TypeScript">
The TypeScript SDK provides dedicated functions for logging media:
- `weave.weaveImage({ data: Buffer })` - for images (PNG format)
- `weave.weaveAudio({ data: Buffer })` - for audio (WAV format)

<Note>

> The TypeScript SDK currently only supports logging images and audio. Video, documents, and HTML logging are available in the Weave Python SDK.

Reviewer: ```suggestion
The TypeScript SDK currently supports logging images and audio. To log videos, documents, or HTML, use the Weave Python SDK instead.
```

</Note>

</Tab>
</Tabs>

> Each media section demonstrates how to log that media type with practical examples.

Reviewer: ```suggestion
The following sections provide practical examples of logging each type of media.
```

> ## Log Images

Reviewer: Nit: Sentence case

> The following examples demonstrate how to generate and log images to Weave's UI.

Reviewer: This seems redundant and it's confusing to have one heading just called Example and another called a really specific thing. Instead, it might be better to have a structure like:

```
## Log images
This section gives basic and advanced examples of drawing and logging images to  Weave using the `Content` annotation.

### Basic: Log an image
<Tabs>

### Advanced: Resize and log a large image
<Tabs>

Author: Good call!

        const data = fs.readFileSync(path);
        return weave.weaveImage({ data });
    });

    // Assuming you have a PNG image file
    await loadImage('sample_image.png');
}

main();
```

Weave logs the image and returns a link to view the image in the terminal.

### Example

The following example logs an image generated via the OpenAI DALL-E API using the `weaveImage` function:

```typescript lines
import {OpenAI} from 'openai';
import * as weave from 'weave';

async function main() {
    await weave.init('image-example');
    const openai = new OpenAI();

    const generateImage = weave.op(async (prompt: string) => {
        const response = await openai.images.generate({
            model: 'dall-e-3',
            prompt: prompt,
            size: '1024x1024',
            quality: 'standard',
            n: 1,
        });
        const imageUrl = response.data[0].url;
        const imgResponse = await fetch(imageUrl);
        const data = Buffer.from(await imgResponse.arrayBuffer());

        return weave.weaveImage({data});
    });

    generateImage('a cat with a pumpkin hat');
}

main();
```

<Note>

> The TypeScript SDK currently only supports PNG images via `weaveImage()`. For other formats, convert to PNG first.

Reviewer 2: This https://github.com/wandb/weave/blob/master/sdks/node/src/media.ts#L29 seem to suggest we do accept a image type . 

Is it not working?

        const data = fs.readFileSync(path);
        return weave.weaveImage({ data });
    });

    // Assuming you have a PNG image file
    await loadImage('sample_image.png');
}

main();
```

Weave logs the image and returns a link to the trace where you can view the image.

### Example

The following example logs an image generated via the OpenAI DALL-E API using the `weaveImage` function:

```typescript lines
import {OpenAI} from 'openai';
import * as weave from 'weave';

async function main() {
    await weave.init('your-team-name/your-project-name');
    const openai = new OpenAI();

    const generateImage = weave.op(async (prompt: string) => {
        const response = await openai.images.generate({
            model: 'dall-e-3',
            prompt: prompt,
            size: '1024x1024',
            quality: 'standard',
            n: 1,
        });
        const imageUrl = response.data[0].url;
        const imgResponse = await fetch(imageUrl);
        const data = Buffer.from(await imgResponse.arrayBuffer());

        return weave.weaveImage({data});
    });

    generateImage('a cat with a pumpkin hat');
}

main();
```

  </Tab>
</Tabs>

## Log Video

> The following examples demonstrate how to generate and log videos to Weave's UI.

Reviewer: "to Weave's UI" seems weird.

video_data = download_youtube_video("https://www.youtube.com/watch?v=dQw4w9WgXcQ")
result = analyze_video(video_data)
```

</Tab>
<Tab title="TypeScript">

> ```plaintext

Reviewer: Why not format this as a callout, and also call it from a snippet to maximize reuse?
